Bodhi is not compatible with click 7.0.0

Quite a few unit tests [fail](https://ci.centos.org/job/bodhi-bodhi-unit-pip/63/console) with click 7.0.0. Some examples:

```
=================================== FAILURES ===================================                                                                                                                                  
__________________ TestEdit.test_wrong_update_title_argument ___________________                
                                                                                                                  
self = <bodhi.tests.client.test___init__.TestEdit testMethod=test_wrong_update_title_argument>    
                                                                                                                            
    def test_wrong_update_title_argument(self):                                                                                                                                                                    
        """                                                                    
             Assert that an error is given if the edit update argument given is not an update id                                           
             nor an update title.                                                                                                                                                                                  
            """                                                                                                                                                                                                    
        runner = testing.CliRunner()                                                                                                                                                                               
                                                                                                                                                                                                                   
        result = runner.invoke(                                                                                                                                                                                   
            client.edit, ['drupal7-i18n-1.17-1', '--user', 'bowlofeggs',                                                                                                                                           
                          '--password', 's3kr3t', '--notes', 'this is an edited note',                                                                                                                            
                          '--url', 'http://localhost:6543'])                                                    
        self.assertEqual(result.exit_code, 2)                                                           
        expected = u'Usage: edit [OPTIONS] UPDATE\n\n' \                                                                                                                                                          
                   u'Error: Invalid value for "update": ' \                                                           
                   u'Please provide an Update ID or an Update Title\n'         
                                                                                                                                                                                                                   >       self.assertEqual(result.output, expected)                                                                                                                                                                 
E       AssertionError: u'Usage: edit [OPTIONS] UPDATE\n\nError: Invalid value for "UPDATE": Please prov [truncated]... != u'Usage: edit [OPTIONS] UPDATE\n\nError: Invalid value for "update": Please prov [trunca
ted]...                                                                                
E         Usage: edit [OPTIONS] UPDATE                                                        
E                                                                                             
E       - Error: Invalid value for "UPDATE": Please provide an Update ID or an Update Title
E       ?                           ^^^^^^                                             
E       + Error: Invalid value for "update": Please provide an Update ID or an Update Title                                                                                                                        
E       ?                           ^^^^^^                                                                                                               
                                                                                               
bodhi/tests/client/test___init__.py:1443: AssertionError
```

```
___________ test_skopeo_copy_cert[missing_key-Cannot find key file] ____________

breakage = 'missing_key', error = 'Cannot find key file'

>   ???

<string>:3: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

breakage = 'missing_key', error = 'Cannot find key file'

    @responses.activate
    @pytest.mark.parametrize(('breakage', 'error'), [
        (None, None),
        ('missing_cert', 'Cannot find certificate file'),
        ('missing_key', 'Cannot find key file'),
        ('missing_cert_and_key', 'Wrong/missing cert'),
    ])
    def test_skopeo_copy_cert(breakage, error):
        """
        Test authentication with a certificate
        """
        runner = testing.CliRunner()
    
        certdir1 = tempfile.mkdtemp()
        certdir2 = tempfile.mkdtemp()
        try:
            certs = {}
            for certdir, reg in ((certdir1, 'registry1.example.com'),
                                 (certdir2, 'registry2.example.com')):
                cert = os.path.join(certdir, reg + '.cert')
                if breakage not in ('missing_cert', 'missing_cert_and_key'):
                    with open(cert, 'w'):
                        pass
                key = os.path.join(certdir, reg + '.key')
                if breakage not in ('missing_key', 'missing_cert_and_key'):
                    with open(key, 'w'):
                        pass
                certs[reg] = (cert, key)
    
            content_type = MEDIA_TYPE_MANIFEST_V2
            reg1 = MockRegistry('registry1.example.com')
            reg2 = MockRegistry('registry2.example.com')
            digest = reg1.add_fake_image('repo1', 'latest', content_type)
    
            with check_certificates(get_cert=certs['registry1.example.com'],
                                    put_cert=certs['registry2.example.com']):
                args = ['--src-cert-dir', certdir1,
                        '--dest-cert-dir', certdir2,
                        'docker://registry1.example.com/repo1:latest',
                        'docker://registry2.example.com/repo2:latest']
    
                if breakage is None:
                    result = runner.invoke(
                        skopeo_lite.copy,
                        args,
                        catch_exceptions=False)
    
                    assert result.exit_code == 0
                    reg2.check_fake_image('repo2', 'latest', digest, content_type)
                else:
                    with pytest.raises(Exception) as e:
                        runner.invoke(
                            skopeo_lite.copy,
                            args,
                            catch_exceptions=False)
>                   assert error in str(e)
E                   assert 'Cannot find key file' in "/usr/lib64/python2.7/site-packages/click/core.py:555: TypeError: copy() got an unexpected keyword argument 'dest_creds'"
E                    +  where "/usr/lib64/python2.7/site-packages/click/core.py:555: TypeError: copy() got an unexpected keyword argument 'dest_creds'" = str(<ExceptionInfo TypeError tblen=5>)

bodhi/tests/server/scripts/test_skopeo_lite.py:593: AssertionError
```

--- a/bodhi/server/__init__.py
+++ b/bodhi/server/__init__.py
@@ -31,15 +31,11 @@ from sqlalchemy import engine_from_config, event
 from sqlalchemy.orm import scoped_session, sessionmaker
 import six
 
-from bodhi.server import bugs, buildsys, ffmarkdown
+from bodhi.server import bugs, buildsys
 from bodhi.server.config import config as bodhi_config
 
 
 log = logging.getLogger(__name__)
-
-
-# TODO -- someday move this externally to "fedora_flavored_markdown"
-ffmarkdown.inject()
 
 
 #

--- a/bodhi/server/ffmarkdown.py
+++ b/bodhi/server/ffmarkdown.py
@@ -1,5 +1,5 @@
 # -*- coding: utf-8 -*-
-# Copyright © 2014-2017 Red Hat, Inc. and others.
+# Copyright © 2014-2018 Red Hat, Inc. and others.
 #
 # This file is part of Bodhi.
 #
@@ -23,10 +23,15 @@ Fedora-flavored Markdown.
 Author: Ralph Bean <rbean@redhat.com>
 """
 
+from markdown.extensions import Extension
 import markdown.inlinepatterns
 import markdown.postprocessors
 import markdown.util
 import pyramid.threadlocal
+
+
+MENTION_RE = r'(?<!\S)(@\w+)'
+BUGZILLA_RE = r'([a-zA-Z]+)(#[0-9]{5,})'
 
 
 def user_url(name):
@@ -72,54 +77,76 @@ def bug_url(tracker, idx):
         return None
 
 
-def inject():
-    """Hack out python-markdown to do the autolinking that we want."""
-    # build some Pattern objects for @mentions, #bugs, etc...
-    class MentionPattern(markdown.inlinepatterns.Pattern):
-        def handleMatch(self, m):
-            el = markdown.util.etree.Element("a")
-            name = markdown.util.AtomicString(m.group(2))
-            el.set('href', user_url(name[1:]))
-            el.text = name
-            return el
+class MentionPattern(markdown.inlinepatterns.Pattern):
+    """Match username mentions and point to their profiles."""
 
-    class BugzillaPattern(markdown.inlinepatterns.Pattern):
-        def handleMatch(self, m):
-            tracker = markdown.util.AtomicString(m.group(2))
-            idx = markdown.util.AtomicString(m.group(3))
-            url = bug_url(tracker, idx[1:])
+    def handleMatch(self, m):
+        """
+        Build and return an Element that links to the matched User's profile.
 
-            if url is None:
-                return tracker + idx
+        Args:
+            m (re.MatchObject): The regex match on the username.
+        Return:
+            xml.etree.Element: An html anchor referencing the user's profile.
+        """
+        el = markdown.util.etree.Element("a")
+        name = markdown.util.AtomicString(m.group(2))
+        el.set('href', user_url(name[1:]))
+        el.text = name
+        return el
 
-            el = markdown.util.etree.Element("a")
-            el.set('href', url)
-            el.text = idx
-            return el
 
-    MENTION_RE = r'(?<!\S)(@\w+)'
-    BUGZILLA_RE = r'([a-zA-Z]+)(#[0-9]{5,})'
+class BugzillaPattern(markdown.inlinepatterns.Pattern):
+    """Match bug tracker patterns."""
 
-    class SurroundProcessor(markdown.postprocessors.Postprocessor):
-        def run(self, text):
-            return "<div class='markdown'>" + text + "</div>"
+    def handleMatch(self, m):
+        """
+        Build and return an Element that links to the referenced bug.
 
-    # monkey-patch the build_inlinepatterns func to insert our patterns
-    original_pattern_builder = markdown.build_inlinepatterns
+        Args:
+            m (re.MatchObject): The regex match on the bug.
+        Returns:
+            xml.etree.Element: An html anchor referencing the matched bug.
+        """
+        tracker = markdown.util.AtomicString(m.group(2))
+        idx = markdown.util.AtomicString(m.group(3))
+        url = bug_url(tracker, idx[1:])
 
-    def extended_pattern_builder(md_instance, **kwargs):
-        patterns = original_pattern_builder(md_instance, **kwargs)
-        patterns['mention'] = MentionPattern(MENTION_RE, md_instance)
-        patterns['bugzillas'] = BugzillaPattern(BUGZILLA_RE, md_instance)
-        return patterns
+        if url is None:
+            return tracker + idx
 
-    markdown.build_inlinepatterns = extended_pattern_builder
+        el = markdown.util.etree.Element("a")
+        el.set('href', url)
+        el.text = idx
+        return el
 
-    original_postprocessor_builder = markdown.build_postprocessors
 
-    def extended_postprocessor_builder(md_instance, **kwargs):
-        processors = original_postprocessor_builder(md_instance, **kwargs)
-        processors['surround'] = SurroundProcessor(md_instance)
-        return processors
+class SurroundProcessor(markdown.postprocessors.Postprocessor):
+    """A postprocessor to surround the text with a markdown <div>."""
 
-    markdown.build_postprocessors = extended_postprocessor_builder
+    def run(self, text):
+        """
+        Return text wrapped in a <div> with a markdown class.
+
+        Args:
+            text (str): The text to wrap in a <div>.
+        Returns:
+            str: The text wrapped in a <div>.
+        """
+        return "<div class='markdown'>" + text + "</div>"
+
+
+class BodhiExtension(Extension):
+    """Bodhi's markdown Extension."""
+
+    def extendMarkdown(self, md, md_globals):
+        """
+        Extend markdown to add our patterns and postprocessor.
+
+        Args:
+            md (Markdown): An instance of the Markdown class.
+            md_globals (dict): Contains all the various global variables within the markdown module.
+        """
+        md.inlinePatterns.add('mention', MentionPattern(MENTION_RE, md), '_end')
+        md.inlinePatterns.add('bugzilla', BugzillaPattern(BUGZILLA_RE, md), '_end')
+        md.postprocessors.add('surrond', SurroundProcessor(md), '_end')

--- a/bodhi/server/scripts/skopeo_lite.py
+++ b/bodhi/server/scripts/skopeo_lite.py
@@ -738,7 +738,7 @@ class Copier(object):
 
 
 @main.command()
-@click.option('--src-creds', '--screds', metavar='USERNAME[:PASSWORD]',
+@click.option('--src-creds', '--screds', 'src_creds', metavar='USERNAME[:PASSWORD]',
               help='Use USERNAME[:PASSWORD] for accessing the source registry')
 @click.option('--src-tls-verify', type=bool, default=True,
               help=('require HTTPS and verify certificates when talking to the '
@@ -746,7 +746,7 @@ class Copier(object):
 @click.option('--src-cert-dir', metavar='PATH',
               help=('use certificates at PATH (*.crt, *.cert, *.key) to connect to the '
                     'source registry'))
-@click.option('--dest-creds', '--dcreds', metavar='USERNAME[:PASSWORD]',
+@click.option('--dest-creds', '--dcreds', 'dest_creds', metavar='USERNAME[:PASSWORD]',
               help='Use USERNAME[:PASSWORD] for accessing the destination registry')
 @click.option('--dest-tls-verify', type=bool, default=True,
               help=('require HTTPS and verify certificates when talking to the '
@@ -756,10 +756,11 @@ class Copier(object):
                     'destination registry'))
 @click.argument('src', metavar='SOURCE-IMAGE-NAME')
 @click.argument('dest', metavar='DEST-IMAGE-NAME')
-def copy(src, dest, screds, src_tls_verify, src_cert_dir, dcreds, dest_tls_verify, dest_cert_dir):
+def copy(src, dest, src_creds, src_tls_verify, src_cert_dir, dest_creds, dest_tls_verify,
+         dest_cert_dir):
     """Copy an image from one location to another."""
-    src = parse_spec(src, screds, src_tls_verify, src_cert_dir)
-    dest = parse_spec(dest, dcreds, dest_tls_verify, dest_cert_dir)
+    src = parse_spec(src, src_creds, src_tls_verify, src_cert_dir)
+    dest = parse_spec(dest, dest_creds, dest_tls_verify, dest_cert_dir)
 
     if src.registry != dest.registry:
         tempdir = tempfile.mkdtemp()

--- a/bodhi/server/util.py
+++ b/bodhi/server/util.py
@@ -46,7 +46,7 @@ from six.moves import map
 from six.moves.urllib.parse import urlencode
 import six
 
-from bodhi.server import log, buildsys, Session
+from bodhi.server import ffmarkdown, log, buildsys, Session
 from bodhi.server.config import config
 from bodhi.server.exceptions import RepodataException
 
@@ -480,7 +480,8 @@ def markup(context, text):
         "a",
     ]
 
-    markdown_text = markdown.markdown(text, extensions=['markdown.extensions.fenced_code'])
+    markdown_text = markdown.markdown(text, extensions=['markdown.extensions.fenced_code',
+                                                        ffmarkdown.BodhiExtension()])
 
     # previously, we linkified text in ffmarkdown.py, but this was causing issues like #1721
     # so now we use the bleach linkifier to do this for us.

--- a/bodhi/tests/client/test___init__.py
+++ b/bodhi/tests/client/test___init__.py
@@ -25,6 +25,7 @@ import unittest
 import copy
 
 from click import testing
+import click
 import fedora.client
 import mock
 import munch
@@ -1435,9 +1436,15 @@ class TestEdit(unittest.TestCase):
                           '--password', 's3kr3t', '--notes', 'this is an edited note',
                           '--url', 'http://localhost:6543'])
         self.assertEqual(result.exit_code, 2)
+        # Click 7.0 capitalizes UPDATE, and < 7 does not.
+        if [int(n) for n in click.__version__.split('.')] < [7, 0]:
+            label = 'update'
+        else:
+            label = 'UPDATE'
         expected = u'Usage: edit [OPTIONS] UPDATE\n\n' \
-                   u'Error: Invalid value for "update": ' \
+                   u'Error: Invalid value for "{}": ' \
                    u'Please provide an Update ID or an Update Title\n'
+        expected = expected.format(label)
 
         self.assertEqual(result.output, expected)
 
@@ -1453,9 +1460,15 @@ class TestEdit(unittest.TestCase):
                           '--password', 's3kr3t', '--notes', 'this is an edited note',
                           '--url', 'http://localhost:6543'])
         self.assertEqual(result.exit_code, 2)
+        # Click 7.0 capitalizes UPDATE, and < 7 does not.
+        if [int(n) for n in click.__version__.split('.')] < [7, 0]:
+            label = 'update'
+        else:
+            label = 'UPDATE'
         expected = u'Usage: edit [OPTIONS] UPDATE\n\n' \
-                   u'Error: Invalid value for "update": ' \
+                   u'Error: Invalid value for "{}": ' \
                    u'Please provide an Update ID or an Update Title\n'
+        expected = expected.format(label)
 
         self.assertEqual(result.output, expected)
 

